Core/Src/main.c

```diff
@@ -0,0 +1,935 @@
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+uint32_t DWT_Delay_Init(void)
+{
+    /* Disable TRC */
+    CoreDebug->DEMCR &= ~CoreDebug_DEMCR_TRCENA_Msk; // ~0x01000000;
+    /* Enable TRC */
+    CoreDebug->DEMCR |=  CoreDebug_DEMCR_TRCENA_Msk; // 0x01000000;
+
+    /* Disable clock cycle counter */
+    DWT->CTRL &= ~DWT_CTRL_CYCCNTENA_Msk; //~0x00000001;
+    /* Enable  clock cycle counter */
+    DWT->CTRL |=  DWT_CTRL_CYCCNTENA_Msk; //0x00000001;
+
+    /* Reset the clock cycle counter value */
+    DWT->CYCCNT = 0;
+
+    /* 3 NO OPERATION instructions */
+    __ASM volatile ("NOP");
+    __ASM volatile ("NOP");
+    __ASM volatile ("NOP");
+
+    /* Check if clock cycle counter has started */
+    if(DWT->CYCCNT)
+    {
+        return 0; /*clock cycle counter started*/
+    }
+    else
+    {
+        return 1; /*clock cycle counter not started*/
+    }
+}
+__STATIC_INLINE void DWT_Delay_us(volatile uint32_t au32_microseconds)
+{
+    uint32_t au32_initial_ticks = DWT->CYCCNT;
+    uint32_t au32_ticks = (HAL_RCC_GetHCLKFreq() / 1000000);
+    au32_microseconds *= au32_ticks;
+    while ((DWT->CYCCNT - au32_initial_ticks) < au32_microseconds-au32_ticks);
+}
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+TIM_HandleTypeDef htim1;
+TIM_HandleTypeDef htim2;
+TIM_HandleTypeDef htim3;
+TIM_HandleTypeDef htim4;
+TIM_HandleTypeDef htim6;
+TIM_HandleTypeDef htim8;
+TIM_HandleTypeDef htim15;
+TIM_HandleTypeDef htim16;
+DMA_HandleTypeDef hdma_tim15_ch1;
+
+/* USER CODE BEGIN PV */
+uint16_t value = TROTTLE_MIN;// throttle value to be sent
+uint16_t ctr = 0;            // arming sequence counter
+uint8_t trottle_down = 0;    // dunnot change trottle up.
+uint8_t armed = 0;           // arming sequence ended
+
+uint32_t motor[DSHOT_FRAME_SIZE];// duty cycles array
+volatile uint16_t DT = 0;
+volatile int8_t motorSpeeds[3];
+volatile int8_t motor_new = 0;
+extern USBD_HandleTypeDef hUsbDeviceFS;
+struct Data {
+    char A, B, C, D;
+};
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_DMA_Init(void);
+static void MX_TIM1_Init(void);
+static void MX_TIM2_Init(void);
+static void MX_TIM16_Init(void);
+static void MX_TIM3_Init(void);
+static void MX_TIM4_Init(void);
+static void MX_TIM8_Init(void);
+static void MX_TIM6_Init(void);
+static void MX_TIM15_Init(void);
+/* USER CODE BEGIN PFP */
+void dshot600(uint32_t *motor, uint16_t value);
+uint16_t prepareDshotPacket(const uint16_t value);
+int16_t enc_val_prev[3];
+struct Data d1 = {1, 9, 3, 0};
+
+float required_speed = 4 ;
+float last_error, error, integral, derevative;
+float Kp=110, Ki=0.1, Kd=0.001;
+
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
+    if (htim == &htim1) {

```

Comment:
Timer 1 is used for encoder. Callbacks would only be called when the counts have overflown. This would not work for motor control.

Timer 6 or some other unused timer should be used for periodic callback.

Reply:
change for tim6 and add implementation